### DISCUSSION TITLE
Components are unstyled in a Webpack Module Federation setup with Shadow DOM and StyleProvider

### DISCUSSION THREAD

**Discussion by bertanyaman** [ant-design/ant-design#51612] [General]
### Reproduction link

[https://github.com/bertanyaman/mf-antd-example](https://github.com/bertanyaman/mf-antd-example)

### Steps to reproduce

Set up a primary React application that uses Webpack Module Federation.
Integrate Ant Design and include an Input component in the primary application.
Mount two remote applications, each also using Ant Design components.
For each remote application, encapsulate its rendered content within the Shadow DOM.
Wrap the Ant Design components within a StyleProvider and set the container prop to target the Shadow DOM container.

### What is expected?

The Ant Design components in both remote applications should render with the expected Ant Design styles, provided by the StyleProvider and scoped to the Shadow DOM.

### What is actually happening?

Styles for the Ant Design components in one of the remote applications are not injected, resulting in unstyled or partially styled components within the Shadow DOM.

| Environment | Info |
| --- | --- |
| antd | 5.21.5 |
| React | 18.2.0 |
| System | Windows 11 |
| Browser | Chrome |

---
This issue occurs specifically in Webpack Module Federation setups where multiple remotes encapsulate components within their own Shadow DOMs. Despite the StyleProvider being used with the container prop pointing to the Shadow DOM, some remote components do not receive the styles correctly. This behavior might stem from conflicts or limitations with how CSS styles are applied or scoped in Shadow DOM when used with federated modules.

**Comment by petercat-assistant** [comment]
It seems you're experiencing an issue with Ant Design components not being styled correctly in a Webpack Module Federation setup when using Shadow DOM. This can be a complex issue due to the interaction between Shadow DOM, CSS scoping, and module federation.

Here are a few suggestions to troubleshoot and potentially resolve the issue:

1. **Ensure Correct Usage of StyleProvider**: Double-check that the `StyleProvider` is correctly wrapping the Ant Design components and that the `container` prop is accurately pointing to the Shadow DOM container.

2. **CSS Isolation**: Shadow DOM is designed to encapsulate styles, which can sometimes lead to issues with global styles not being applied. Ensure that any global styles needed by Ant Design are correctly scoped or imported within the Shadow DOM.

3. **Check for Conflicts**: Since you're using Webpack Module Federation, ensure there are no conflicts between the styles of the host and remote applications. Conflicting styles or duplicate imports might lead to unexpected behavior.

4. **Verify Module Federation Configuration**: Ensure that the Webpack Module Federation setup is correctly configured to share dependencies like React and Ant Design. Misconfiguration might lead to multiple instances of React or Ant Design being loaded, causing styling issues.

5. **Debugging Styles**: Use browser developer tools to inspect the styles applied to the components. This can help identify if styles are being overridden or not applied at all.

6. **Consult Documentation and Community**: Sometimes, specific configurations or updates in libraries might cause issues. Check the Ant Design and Webpack documentation for any notes on compatibility with Shadow DOM and Module Federation. Additionally, consider reaching out to community forums or GitHub issues for similar cases.

If these steps do not resolve the issue, it might be beneficial to create a minimal reproducible example and share it with the community for more targeted assistance.

For further assistance, please describe your question in the comments and @petercat-assistant to start a conversation with me.

**Comment by zombieJ** [comment]
Since it's not a antd bug but external usage question. Let me move from issue to discussion instead.

**Comment by petercat-assistant** [comment]
It seems like you're experiencing an issue with Ant Design components not being styled correctly when using Webpack Module Federation with Shadow DOM and StyleProvider. This is a known challenge when dealing with Shadow DOM and federated modules, as styles may not be scoped correctly.

Here are a few suggestions to help resolve this issue:

1. **Ensure Correct Container Prop**: Double-check that the `container` prop in `StyleProvider` is correctly pointing to the Shadow DOM container. This is crucial for scoping styles properly.

2. **Scoped Styles**: Make sure that the styles are being scoped correctly within the Shadow DOM. You might need to manually inject styles or use a workaround to ensure styles are applied within the Shadow DOM.

3. **Check for Conflicts**: Look for any potential conflicts or overrides in your CSS that might be affecting the styles. This can sometimes happen with federated modules if styles are not isolated properly.

4. **Update Dependencies**: Ensure that all your dependencies, including Ant Design and Webpack, are up to date. Sometimes, issues like these are resolved in newer versions.

5. **Custom Style Injection**: As a workaround, you might need to manually inject styles into the Shadow DOM. This can be done by creating a style element and appending it to the Shadow DOM.

If these suggestions do not resolve the issue, it might be beneficial to look into community discussions or open an issue on the Ant Design GitHub repository for further assistance.

For further assistance, please describe your question in the comments and @petercat-assistant to start a conversation with me.

**Comment by bertanyaman** [comment]
The issue arises due to sharing the antd library as a shared module in your Webpack Module Federation configuration. When antd is shared, multiple federated modules attempt to use the same instance of antd, leading to conflicts in style injection. Since each module uses its own Shadow DOM and relies on StyleProvider for scoped styling, the shared antd instance can confuse which styles belong to which module.

To resolve this, avoid sharing the antd library in your Webpack Module Federation configuration. Instead, ensure each remote application has its own isolated instance of antd. This guarantees that styles are correctly scoped and injected within the respective Shadow DOM container.
